src/ui/speedscope.rs

```diff
@@ -0,0 +1,168 @@
+use std::collections::{HashMap};
+use std::io;
+use std::io::Write;
+use std::fs::File;
+
+use core::types::StackFrame;
+
+use failure::{Error};
+use serde_json;
+
+/*
+ * This file contains code to export rbspy profiles for use in https://speedscope.app
+ *
+ * The speedscope file format is specified via a JSON schema.
+ * The latest schema can be found here: https://speedscope.app/file-format-schema.json
+ *
+ * This JSON schema conveniently allows to generate type bindings for generating JSON.
+ * You can use https://app.quicktype.io/ to generate serde_json Rust bindings for the
+ * given JSON schema.
+ *
+ * There are multiple variants of the file format. The variant we're going to generate
+ * is the "type: sampled" profile, since it most closely maps to rbspy's data recording
+ * structure.
+ */
+
+#[derive(Debug, Serialize, Deserialize)]
+struct SpeedscopeFile {
+    #[serde(rename = "$schema")]
+    schema: String,
+    profiles: Vec<Profile>,
+    shared: Shared,
+    version: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Profile {
+    #[serde(rename = "type")]
+    profile_type: ProfileType,
+
+    name: String,
+    unit: ValueUnit,
+
+    #[serde(rename = "startValue")]
+    start_value: f64,
+
+    #[serde(rename = "endValue")]
+    end_value: f64,
+
+    samples: Vec<Vec<usize>>,
+    weights: Vec<f64>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Shared {
+    frames: Vec<Frame>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct Frame {
+    name: String,
+    file: Option<String>,
+    line: Option<u32>,
+    col: Option<u32>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+enum ProfileType {
+    #[serde(rename = "evented")]
+    Evented,
+    #[serde(rename = "sampled")]
+    Sampled,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+enum ValueUnit {
+    #[serde(rename = "bytes")]
+    Bytes,
+    #[serde(rename = "microseconds")]
+    Microseconds,
+    #[serde(rename = "milliseconds")]
+    Milliseconds,
+    #[serde(rename = "nanoseconds")]
+    Nanoseconds,
+    #[serde(rename = "none")]
+    None,
+    #[serde(rename = "seconds")]
+    Seconds,
+}
+
+impl SpeedscopeFile {
+  pub fn new(samples: Vec<Vec<usize>>, frames: Vec<Frame>) -> SpeedscopeFile {
+    let end_value = samples.len().clone();
+
+    let weights: Vec<f64> = (&samples).into_iter().map(|_s| 1 as f64).collect();
+
+    SpeedscopeFile {
+      // This is always the same
+      schema: "https://www.speedscope.app/file-format-schema.json".to_string(),
```

Comment:
Could we download the file and vendor it locally?

Reply:
This is part of how speedscope detects that the file is a speedscope file. rbspy doesn't actually need to download this file or do anything fancy with it -- it can be effectively treated as an opaque string.